### PR TITLE
Digging_to_kernel

### PR DESCRIPTION
**Pull request by wumingzhilian**
Digging_to_kernel

### CODE REVIEW

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md [hunk 1]
```diff
@@ -0,0 +1,1114 @@
+# 2022-RealWorld-CTF(experience competition)-digger_into_kernel
```
Comment (mudongliang) [inline _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md]: Refer to [CISCN2017 - babydriver](https://github.com/mudongliang/mudongliang.github.io/edit/master/_posts/2022-01-18-ciscn2017---babydriver.md) and modify the titles and subtitles, e.g., "# -> ##, ##->###". And then add the first ten lines of the referred post here.

Comment (wumingzhilian) [reply _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md]: Thank you very much for your reply, the article has been revised.

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md [hunk 2]
```diff
@@ -0,0 +1,1128 @@
...
+	-cpu kvm64,+smep,+smap \
+	--nographic
+
+```
+
+ `run.sh`  is a qemu script; meanwhile,`rootfs.cpio` is the root filesystem,`bzImage` is the Linux kernel,`-append` enables kalsr and `+smep +smap`enables two protections, one is to prohibit `ret2user`, and the other is to prohibit ROP chain.
+
+To facilitate debugging, we can use the gdb parameter, `-gdb tcp::1234` ,to open the debugging port.
+
+This `run.sh ` scripts don't contain the argument of `-monitor null` , so you can enter the monitor mode by pressing `ctrl+A` and then entering C (Ps: the command execution of the host can be equivalent to escape)
```
Comment (mudongliang) [inline _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md]: This `run.sh` script does not ...

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md [hunk 3]
```diff
@@ -0,0 +1,1128 @@
...
+if ( buf && len <= 0x50 && index <= 0x70 )
+    {
+      copy_to_user(user_data, (char *)buf + (int)index);
+      return 0LL;
+    }
+```
+
+#### Trick
+
+https://elixir.bootlin.com/linux/v5.4.38/source/include/linux/sched.h#L878
```
Comment (mudongliang) [inline _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md:327]: <https://elixir.bootlin.com/linux/v5.4.38/source/include/linux/sched.h#L878>

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md [hunk 4]
```diff
@@ -0,0 +1,1128 @@
...
+
+	extern struct cred *prepare_creds(void);
+	……
+```
+
+Analysis of `prepare_creds`:
+
+This structure is used to apply and generate the cred structure.
+
+https://elixir.bootlin.com/linux/v5.4.38/source/kernel/cred.c#L250
```
Comment (mudongliang) [inline _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md:406]: <https://elixir.bootlin.com/linux/v5.4.38/source/kernel/cred.c#L250>

And the same operation to the single url.

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md [hunk 5]
```diff
@@ -0,0 +1,1128 @@
...
+	kernel_cap_t	cap_effective;	/* caps we can actually use */
+	kernel_cap_t	cap_bset;	/* capability bounding set */
+	kernel_cap_t	cap_ambient;	/* Ambient capability set */
+	……
+
+	extern struct cred *prepare_creds(void);
+	……
+```
+
+Analysis of `prepare_creds`:
```
Comment (mudongliang) [inline _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md:402]: You need to add analysis of the function [cred_init](https://elixir.bootlin.com/linux/latest/source/kernel/cred.c#L689). This helps you verify the length of this dedicated slab.

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md (hunk 2 above)
Comment (wumingzhilian) [reply _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md]: Thank you for your correction!

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md [hunk 6]
```diff
@@ -0,0 +1,1147 @@
...
+	/* Objective and real subjective task credentials (COW): */
+	const struct cred __rcu		*real_cred;
+
+	/* Effective (overridable) subjective task credentials (COW): */
+	const struct cred __rcu		*cred;
+```
+
+The cred structure contains the `permission information` of each process,such as `uid`, `gid`,`suid`, etc.
+
+https://elixir.bootlin.com/linux/v5.4.38/source/include/linux/cred.h#L111
```
Comment (wumingzhilian) [inline _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md:349]: https://elixir.bootlin.com/linux/v5.4.38/source/include/linux/cred.h#L111

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md [hunk 7]
```diff
@@ -0,0 +1,1147 @@
...
+	kernel_cap_t	cap_ambient;	/* Ambient capability set */
+	……
+
+	extern struct cred *prepare_creds(void);
+	……
+```
+
+Analysis of `cred_init` :
+
+https://elixir.bootlin.com/linux/v5.4.38/source/kernel/cred.c#L656
```
Comment (wumingzhilian) [inline _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md:383]: https://elixir.bootlin.com/linux/v5.4.38/source/kernel/cred.c#L656

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md (hunk 1 above)
Comment (wumingzhilian) [reply _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md]: References：
[1] https://mp.weixin.qq.com/s/v9_C43Qh9qXr9tYojrZKgQ
[2] https://xz.aliyun.com/t/11053
[3] https://mudongliang.github.io/2022/01/18/ciscn2017-babydriver.html
[4]https://github.com/g0dA/linuxStack/blob/master/%E5%85%B3%E4%BA%8E%E5%86%85%E5%AD%98%E5%88%86%E9%85%8D%E7%9A%84%E4%B8%80%E4%BA%9B%E4%BA%8B.md
[5] 段鑫峰. 面向Linux内核空间的内存分配隔离方法的研究与实现[D].北京交通大学,2021.DOI:10.26944/d.cnki.gbfju.2021.000664.
[6] https://www.anquanke.com/post/id/259252#h2-6
[7] https://zhuanlan.zhihu.com/p/490588193
[8] https://kernel.blog.csdn.net/article/details/52705552
[9]https://www.jeanleo.com/2018/09/07/%E3%80%90linux%E5%86%85%E5%AD%98%E6%BA%90%E7%A0%81%E5%88%86%E6%9E%90%E3%80%91slub%E5%88%86%E9%85%8D%E7%AE%97%E6%B3%95%EF%BC%883%EF%BC%89/

File: _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md (hunk 5 above)
Comment (wumingzhilian) [reply _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md:402]: Thank you very much for your correction, the analysis of the cred_init function has been added.

Comment (wumingzhilian) [reply _posts/2022-06-07-RealWorldCTF-digging_into_kernel.md:402]: Thank you very much for your corrections, which have added analysis of cred_init functions.

**Pull request merged**